### PR TITLE
llvm@4: remove libcxx and change cmake option for compiler-rt

### PR DESCRIPTION
**Pull request by Char-Aznable**
Address issue #10371

- [x] Have you followed the [guidelines for contributing](https://github.com/Homebrew/linuxbrew-core/blob/master/CONTRIBUTING.md)?
- [x] Have you checked that there aren't other open [pull requests](https://github.com/Homebrew/linuxbrew-core/pulls) for the same formula update/change?
- [x] Have you built your formula locally with `brew install --build-from-source <formula>`, where `<formula>` is the name of the formula you're submitting?
- [x] Is your test running fine `brew test <formula>`, where `<formula>` is the name of the formula you're submitting?
- [x] Does your build pass `brew audit --strict <formula>` (after doing `brew install <formula>`)?
- [ ] Have you included the output of `brew gist-logs <formula>` of the build failure if your PR fixes a build failure. Please quote the exact error message.

-----

### CODE REVIEW

File: Formula/llvm@4.rb [hunk 1]
```diff
@@ -5,7 +5,7 @@ class LlvmAT4 < Formula
   homepage "https://llvm.org/"
   url "https://releases.llvm.org/4.0.1/llvm-4.0.1.src.tar.xz"
   sha256 "da783db1f82d516791179fe103c71706046561f7972b18f0049242dee6712b51"
-  revision 1
+  revision 2
```
Comment (iMichka) [inline Formula/llvm@4.rb]: revision OS.mac? ? 1 : 2

**Comment by iMichka** [comment]
I made similar updates to `llvm@7` lately: #12459
Can you have a look at what is missing? I think we can:
- comment out some tests. 
- add the simple test that was added `llvm@7`
- add the missing binary stripping
- add the following patch: https://github.com/Homebrew/linuxbrew-core/pull/10389/files#diff-ff9dfeb4cdef4e38fa41da9befe6d095R61

**Comment by iMichka** [comment]
And by the way, do not worry if the CI fails, it may run out of memory. Once this builds fine for you locally, I will run the build on my side, and upload the bottles manually.

**Comment by Char-Aznable** [comment]
@iMichka This builds fine for me locally on centos 6.9:
```
updading 1 outdated package:                                                                                                                     [30/2126]
llvm@4 4.0.1 -> 4.0.1_2                                                                                                                                       
==> Upgrading llvm@4                                                                                                                                          
==> Downloading https://releases.llvm.org/4.0.1/llvm-4.0.1.src.tar.xz                                                                                         
Already downloaded: /home/aznb/.cache/downloads/197abd9f40c3abfd249b718a7d43535106b422dfc9e95e5a207aee7421cb63c4--llvm-4.0.1.src.tar.xz    
==> Downloading https://releases.llvm.org/4.0.1/cfe-4.0.1.src.tar.xz                                                                                          
Already downloaded: /home/aznb/.cache/downloads/53db026ec220cc9dec2bb9be168f7b629ec16e63f8be9517a3dad609bf779664--cfe-4.0.1.src.tar.xz     
==> Downloading https://releases.llvm.org/4.0.1/clang-tools-extra-4.0.1.src.tar.xz                                                                            
Already downloaded: /home/aznb/.cache/downloads/6b5e4eee7ab0dbdcb890587274514ffe0a9256bbb522b73e77916ea5493c87c4--clang-tools-extra-4.0.1.sr
c.tar.xz                                                                                                                                                      
==> Downloading https://releases.llvm.org/4.0.1/openmp-4.0.1.src.tar.xz                                                                                       
Already downloaded: /home/aznb/.cache/downloads/3066319da4994fec869e4c73859fc39a4e0134214c8a32277d09363620ad7ccb--openmp-4.0.1.src.tar.xz  
==> Downloading https://releases.llvm.org/4.0.1/libunwind-4.0.1.src.tar.xz                                                                                    
Already downloaded: /home/aznb/.cache/downloads/2741d491a4885ba842c815196fa42b27a9e4d78e2adb72f9ec5e95cdde00ba26--libunwind-4.0.1.src.tar.xz
==> Downloading https://releases.llvm.org/4.0.1/lld-4.0.1.src.tar.xz
Already downloaded: /home/aznb/.cache/downloads/19608aa1827b39ac62cb68dc5bd2c0cd548d5e0f8e769aeed85f53355fe70158--lld-4.0.1.src.tar.xz     
==> Downloading https://releases.llvm.org/4.0.1/polly-4.0.1.src.tar.xz
Already downloaded: /home/aznb/.cache/downloads/6f7095f47d8e41aa46349656256a47c72acd04815915a54b7a5119c522c74802--polly-4.0.1.src.tar.xz   
==> Downloading https://releases.llvm.org/4.0.1/compiler-rt-4.0.1.src.tar.xz
Already downloaded: /home/aznb/.cache/downloads/b35110d7b7043546fb86af627fd46618bab6173d9e375a2c376e8e5ca046ef58--compiler-rt-4.0.1.src.tar.
xz
==> cmake -G Unix Makefiles .. -DCMAKE_C_FLAGS_RELEASE=-DNDEBUG -DCMAKE_CXX_FLAGS_RELEASE=-DNDEBUG -DCMAKE_INSTALL_PREFIX=/home/aznb/.linuxb
rew/Cellar/llvm@4/4.0.1_2 -DCMAK
==> make install
==> strip --strip-unneeded --preserve-dates /home/aznb/.linuxbrew/Cellar/llvm@4/4.0.1_2/bin/llvm-nm /
==> Caveats
To use the bundled libc++ please add the following LDFLAGS:
  LDFLAGS="-L/home/aznb/.linuxbrew/opt/llvm@4/lib"                      

llvm@4 is keg-only, which means it was not symlinked into /home/aznb/.linuxbrew,                                                           
because this is an alternate version of another formula.

If you need to have llvm@4 first in your PATH run:
  echo 'export PATH="/home/aznb/.linuxbrew/opt/llvm@4/bin:$PATH"' >> ~/.bash_profile                                                       

For compilers to find llvm@4 you may need to set:
  export LDFLAGS="-L/home/aznb/.linuxbrew/opt/llvm@4/lib"
  export CPPFLAGS="-I/home/aznb/.linuxbrew/opt/llvm@4/include"

==> Summary
🍺  /home/aznb/.linuxbrew/Cellar/llvm@4/4.0.1_2: 2,108 files, 3.4GB, built in 23 minutes 15 seconds
```

**Comment by Char-Aznable** [comment]
Update the tests and they all pass on my machine

**Comment by Char-Aznable** [comment]
@iMichka Any luck of reproducing the build on your machine?

**Comment by iMichka** [comment]
I will be travelling this weekend. I'll have a look at this when I come back next week.

File: Formula/llvm@4.rb [hunk 2]
```diff
@@ -242,6 +254,15 @@ def caveats; <<~EOS
       }
     EOS
 
+    unless OS.mac?
```
Comment (iMichka) [inline Formula/llvm@4.rb]: This test can be simplified and aligned with the llvm formula:
```
    unless OS.mac?
      system "#{bin}/clang++", "-v", "test.cpp", "-o", "test"
      assert_equal "Hello World!", shell_output("./test").chomp
    end
```

File: Formula/llvm@4.rb [hunk 3]
```diff
@@ -207,11 +218,12 @@ def caveats; <<~EOS
       }
     EOS
 
-    system "#{bin}/clang", "-L#{lib}", "-fopenmp", "-nobuiltininc",
```
Comment (iMichka) [inline Formula/llvm@4.rb]: Please align this with the llvm@4 formula, the code should be:
```
    system "#{bin}/clang", "-L#{lib}", "-fopenmp", "-nobuiltininc",
                           "-I#{lib}/clang/#{clean_version}/include",
                           "omptest.c", "-o", "omptest", *ENV["LDFLAGS"].split
    testresult = shell_output(testpath/"omptest")
```

**Comment by iMichka** [comment]
@Char-Aznable I think this looks good. I made two minor comments for the tests. Can you also rebase this on master?

**Comment by Char-Aznable** [comment]
@iMichka Those test codes you suggest don't work on my machine because the compiled test binaries have broken linkage.

File: Formula/llvm@4.rb (hunk 3 above)
Comment (Char-Aznable) [reply Formula/llvm@4.rb]: This gives error:
```
Error: llvm@4: failed
An exception occurred within a child process:
  NameError: undefined local variable or method `clean_version' for #<Formulary::FormulaNamespace7b2fdd774dabbd0caf9d5bc8d051045d::LlvmAT4:0x0000000003607838>                                
Did you mean?  linked_version
/home/aznb/.linuxbrew/Homebrew/Library/Taps/homebrew/homebrew-core/Formula/llvm@4.rb:222:in `block in <class:LlvmAT4>'                                                     
/home/aznb/.linuxbrew/Homebrew/Library/Homebrew/formula.rb:1727:in `block (3 levels) in run_test'                                                                          
/home/aznb/.linuxbrew/Homebrew/Library/Homebrew/utils.rb:478:in `with_env'
/home/aznb/.linuxbrew/Homebrew/Library/Homebrew/formula.rb:1726:in `block (2 levels) in run_test'                                                                          
/home/aznb/.linuxbrew/Homebrew/Library/Homebrew/formula.rb:866:in `with_logging'                                                                                           
/home/aznb/.linuxbrew/Homebrew/Library/Homebrew/formula.rb:1725:in `block in run_test'                                                                                     
/home/aznb/.linuxbrew/Homebrew/Library/Homebrew/formula.rb:1964:in `block in mktemp'                                                                                       
/home/aznb/.linuxbrew/Homebrew/Library/Homebrew/mktemp.rb:57:in `block in run'                                                                                             
/home/aznb/.linuxbrew/Homebrew/Library/Homebrew/mktemp.rb:57:in `chdir'
/home/aznb/.linuxbrew/Homebrew/Library/Homebrew/mktemp.rb:57:in `run'
/home/aznb/.linuxbrew/Homebrew/Library/Homebrew/formula.rb:1963:in `mktemp'
/home/aznb/.linuxbrew/Homebrew/Library/Homebrew/formula.rb:1719:in `run_test'                                                                                              
/home/aznb/.linuxbrew/Homebrew/Library/Homebrew/test.rb:30:in `block in <main>'                                                                                            
/home/aznb/.linuxbrew/Homebrew/Library/Homebrew/vendor/portable-ruby/2.3.7/lib/ruby/2.3.0/timeout.rb:91:in `block in timeout'                                              
/home/aznb/.linuxbrew/Homebrew/Library/Homebrew/vendor/portable-ruby/2.3.7/lib/ruby/2.3.0/timeout.rb:33:in `block in catch'                                                
/home/aznb/.linuxbrew/Homebrew/Library/Homebrew/vendor/portable-ruby/2.3.7/lib/ruby/2.3.0/timeout.rb:33:in `catch'                                                         
/home/aznb/.linuxbrew/Homebrew/Library/Homebrew/vendor/portable-ruby/2.3.7/lib/ruby/2.3.0/timeout.rb:33:in `catch'                                                         
/home/aznb/.linuxbrew/Homebrew/Library/Homebrew/vendor/portable-ruby/2.3.7/lib/ruby/2.3.0/timeout.rb:106:in `timeout'                                                      
/home/aznb/.linuxbrew/Homebrew/Library/Homebrew/test.rb:29:in `<main>'

```

File: Formula/llvm@4.rb (hunk 2 above)
Comment (Char-Aznable) [reply Formula/llvm@4.rb]: This gives error:
```
Testing llvm@4
==> /home/aznb/.linuxbrew/Cellar/llvm@4/4.0.1_2/bin/llvm-config --prefix
==> /home/aznb/.linuxbrew/Cellar/llvm@4/4.0.1_2/bin/clang -L/home/aznb/.linuxbrew/lib -L/home/aznb/.linuxbrew/Cellar/llvm@4/4.0.1_2/lib -fopenmp -nobuiltininc -I/home/aznb/.linuxbrew/Cellar/llvm@4/4.0.1_2/lib/clang/4.0.1/include -Wl,--dynamic-linker=/home/aznb/.linuxbrew/lib/ld.so -Wl,-rpath=/home/aznb/.linuxbrew/Cellar/llvm@4/4.0.1_2/lib omptest.c -o omptest
==> /tmp/llvm@4-test-20190424-4576-zwgdwz/omptest
==> /home/aznb/.linuxbrew/Cellar/llvm@4/4.0.1_2/bin/clang++ -v test.cpp -o test
==> ./test
./test: : ELF: x: Error 18446744073112320375
pgrep: invalid option -- 'q'

Usage:
 pgrep [options] <pattern>

Options:
 -d, --delimiter <string>  specify output delimiter
 -l, --list-name           list PID and process name
 -a, --list-full           list PID and full command line
 -v, --inverse             negates the matching
 -w, --lightweight         list all TID
 -c, --count               count of matching processes
 -f, --full                use full process name to match
 -g, --pgroup <PGID,...>   match listed process group IDs
 -G, --group <GID,...>     match real group IDs
 -n, --newest              select most recently started
 -o, --oldest              select least recently started
 -P, --parent <PPID,...>   match only child processes of the given parent
 -s, --session <SID,...>   match session IDs
 -t, --terminal <tty,...>  match by controlling terminal
 -u, --euid <ID,...>       match by effective IDs
 -U, --uid <ID,...>        match by real IDs
 -x, --exact               match exactly with the command name
 -F, --pidfile <file>      read PIDs from file
 -L, --logpidfile          fail if PID file is not locked
 --ns <PID>                match the processes that belong to the same
                           namespace as <pid>
 --nslist <ns,...>         list which namespaces will be considered for
                           the --ns option.
                           Available namespaces: ipc, mnt, net, pid, user, uts

 -h, --help     display this help and exit
 -V, --version  output version information and exit

For more details see pgrep(1).
Error: llvm@4: failed
An exception occurred within a child process:
  Test::Unit::AssertionFailedError: <0> expected but was
<127>.
/home/aznb/.linuxbrew/Homebrew/Library/Homebrew/vendor/portable-ruby/2.3.7/lib/ruby/gems/2.3.0/gems/test-unit-3.1.5/lib/test/unit/assertions.rb:55:in `block in assert_block'
/home/aznb/.linuxbrew/Homebrew/Library/Homebrew/vendor/portable-ruby/2.3.7/lib/ruby/gems/2.3.0/gems/test-unit-3.1.5/lib/test/unit/assertions.rb:1632:in `_wrap_assertion'
/home/aznb/.linuxbrew/Homebrew/Library/Homebrew/vendor/portable-ruby/2.3.7/lib/ruby/gems/2.3.0/gems/test-unit-3.1.5/lib/test/unit/assertions.rb:53:in `assert_block'
/home/aznb/.linuxbrew/Homebrew/Library/Homebrew/vendor/portable-ruby/2.3.7/lib/ruby/gems/2.3.0/gems/test-unit-3.1.5/lib/test/unit/assertions.rb:240:in `assert_equal'
/home/aznb/.linuxbrew/Homebrew/Library/Homebrew/formula_assertions.rb:13:in `shell_output'
/home/aznb/.linuxbrew/Homebrew/Library/Taps/homebrew/homebrew-core/Formula/llvm@4.rb:259:in `block in <class:LlvmAT4>'
/home/aznb/.linuxbrew/Homebrew/Library/Homebrew/formula.rb:1727:in `block (3 levels) in run_test'
/home/aznb/.linuxbrew/Homebrew/Library/Homebrew/utils.rb:478:in `with_env'
/home/aznb/.linuxbrew/Homebrew/Library/Homebrew/formula.rb:1726:in `block (2 levels) in run_test'
/home/aznb/.linuxbrew/Homebrew/Library/Homebrew/formula.rb:866:in `with_logging'
/home/aznb/.linuxbrew/Homebrew/Library/Homebrew/formula.rb:1725:in `block in run_test'
/home/aznb/.linuxbrew/Homebrew/Library/Homebrew/formula.rb:1964:in `block in mktemp'
/home/aznb/.linuxbrew/Homebrew/Library/Homebrew/mktemp.rb:57:in `block in run'
/home/aznb/.linuxbrew/Homebrew/Library/Homebrew/mktemp.rb:57:in `chdir'
/home/aznb/.linuxbrew/Homebrew/Library/Homebrew/mktemp.rb:57:in `run'
/home/aznb/.linuxbrew/Homebrew/Library/Homebrew/formula.rb:1963:in `mktemp'
/home/aznb/.linuxbrew/Homebrew/Library/Homebrew/formula.rb:1719:in `run_test'
/home/aznb/.linuxbrew/Homebrew/Library/Homebrew/test.rb:30:in `block in <main>'
/home/aznb/.linuxbrew/Homebrew/Library/Homebrew/vendor/portable-ruby/2.3.7/lib/ruby/2.3.0/timeout.rb:91:in `block in timeout'
/home/aznb/.linuxbrew/Homebrew/Library/Homebrew/vendor/portable-ruby/2.3.7/lib/ruby/2.3.0/timeout.rb:33:in `block in catch'
/home/aznb/.linuxbrew/Homebrew/Library/Homebrew/vendor/portable-ruby/2.3.7/lib/ruby/2.3.0/timeout.rb:33:in `catch'
/home/aznb/.linuxbrew/Homebrew/Library/Homebrew/vendor/portable-ruby/2.3.7/lib/ruby/2.3.0/timeout.rb:33:in `catch'
/home/aznb/.linuxbrew/Homebrew/Library/Homebrew/vendor/portable-ruby/2.3.7/lib/ruby/2.3.0/timeout.rb:106:in `timeout'
/home/aznb/.linuxbrew/Homebrew/Library/Homebrew/test.rb:29:in `<main>'

```

**Comment by Char-Aznable** [comment]
@iMichka The merge have too many conflicts that I don't want to waste my time to resolve. If you are still interested, I can open a new pr. But I'm not going to change the test code you suggest because it doesn't work.

**Comment by iMichka** [comment]
Hi. Can you just open a new PR? I will have a look the tests. We are really close to get this done.

**Pull request closed without merging**